tusclient/request.py

```diff
@@ -42,40 +42,49 @@ def __init__(self, uploader):
         self.response_headers = {}
         self.status_code = None
         self.response_content = None
-        self.file = uploader.get_file_stream()
-        self.file.seek(uploader.offset)
 
         self._request_headers = {
             'upload-offset': str(uploader.offset),
             'Content-Type': 'application/offset+octet-stream'
         }
         self._request_headers.update(uploader.get_headers())
-        self._content_length = uploader.get_request_length()
         self._upload_checksum = uploader.upload_checksum
         self._checksum_algorithm = uploader.checksum_algorithm
         self._checksum_algorithm_name = uploader.checksum_algorithm_name
 
-    def add_checksum(self, chunk: bytes):
-        if self._upload_checksum:
-            self._request_headers['upload-checksum'] = \
-                ' '.join((
-                    self._checksum_algorithm_name,
-                    base64.b64encode(
-                        self._checksum_algorithm(chunk).digest()
-                    ).decode('ascii'),
-                ))
+        self._chunk = ChunkReader(
+            uploader.get_file_stream(),
+            uploader.offset,
+            uploader.get_request_length(),
+        )
+
+    def add_checksum(self, file):
+        checksum = self._checksum_algorithm()
+        chunk = file.read(8192)
+        while chunk:
+            checksum.update(chunk)
+            chunk = file.read(8192)
+
+        self._request_headers['upload-checksum'] = \
+            ' '.join((
+                self._checksum_algorithm_name,
+                base64.b64encode(
+                    checksum.digest()
+                ).decode('ascii'),
+            ))
 
 
 class TusRequest(BaseTusRequest):
     """Class to handle async Tus upload requests"""
+
     def perform(self):
         """
         Perform actual request.
         """
+        if self._upload_checksum:
+            self.add_checksum(self._chunk.reset())
         try:
-            chunk = self.file.read(self._content_length)
```

Comment:
Doesn't this mean that he file is already read in chunks?

Reply:
Yes, each tus-chunk will be completely loaded in memory (I shouldn’t have used the word “chunk” in my initial comment).

Still if your chunk is 100 MB big, it is suboptimal to load those 100MB in memory…

Reply:
I see, it's only the chunks that are fully read (and which this PR tries to avoid). Thanks for clarifying.